invenio_vocabularies/datastreams/readers.py

```diff
@@ -8,41 +8,54 @@
 
 """Readers module."""
 
+import csv
+import json
 import re
 import tarfile
+import zipfile
 from abc import ABC, abstractmethod
+from json.decoder import JSONDecodeError
 
 import requests
 import yaml
 
-from .datastreams import StreamEntry
+from .errors import ReaderError
 
 
 class BaseReader(ABC):
     """Base reader."""
 
-    def __init__(self, origin, *args, **kwargs):
-        """Constructor."""
+    def __init__(self, origin=None, mode='r', *args, **kwargs):
+        """Constructor.
+
+        :param origin: Data source (e.g. filepath).
+                       Can be none in case of piped readers.
+        """
         self._origin = origin
+        self._mode = mode
 
     @abstractmethod
-    def read(self, *args, **kwargs):
-        """Reads the content from the origin.
-
-        Yields `StreamEntry` objects.
-        """
+    def _read(self, fp, *args, **kwargs):
```

Comment:
or `iterate`? i.e. `read` opens/reads but this function is the one actually looping... 🤔

Reply:
Agree, any of the below would be possible:

-  `_iter`
- `_iterate`
- `iter`

I somewhat dislike that this starts looking very much like a Python iterator class... Though our use-case does differ because we do the piping etc.

Reply:
done, choose `_iter`. Because it is an iterator + its internal, should not be called from outside `read`.